### ISSUE TITLE
Memory leak?

### ISSUE THREAD

**Issue by barrucadu** [barrucadu/irc-conduit#4]
Email:

> Hello,
> 
> I'm using your library irc-client in my new project for monitoring and storing chat data in some high traffic channels.
> 
> First, let me thank you for you work, as I find your library best out of all available options.
> 
> Now to the problem. I noticed that my program eats up a lot memory with time, so I tried to check for any space leaks. I did a heap profile of my program running for about a day and the graph shows that there might be a problem in module Network.IRC.Conduit.
> 
> I attached the graph in this mail. I understand that it most likely not very helpful to pinpoint the problem, but I don't know how to get more detailed data.
> 
> Could you, please, help me with this? Let me know if you need any additional info.

Graph: https://misc.barrucadu.co.uk/gh-issue-files/2016-08-12-irc-conduit-memory-leak.pdf

**Comment by bazukas** [comment]
Hello, have you been able to replicate? Saw this today, maybe this is connected?
http://www.well-typed.com/blog/2016/09/sharing-conduit/

**Comment by barrucadu** [comment]
I ran a simple IRC client that just connected to freenode and sat in #freenode for a couple of weeks, with a load of profiling options: [hb](https://github.com/barrucadu/irc-conduit/files/531444/2016-10-15-irc-conduit-memory-leak-hb.pdf), [hc](https://github.com/barrucadu/irc-conduit/files/531439/2016-10-15-irc-conduit-memory-leak-hc.pdf), [hd](https://github.com/barrucadu/irc-conduit/files/531442/2016-10-15-irc-conduit-memory-leak-hd.pdf), [hm](https://github.com/barrucadu/irc-conduit/files/531443/2016-10-15-irc-conduit-memory-leak-hm.pdf), [hr](https://github.com/barrucadu/irc-conduit/files/531441/2016-10-15-irc-conduit-memory-leak-hr.pdf), [hy](https://github.com/barrucadu/irc-conduit/files/531440/2016-10-15-irc-conduit-memory-leak-hy.pdf). It definitely looks like a memory leak.

That blog post looks like it might be related, as it looks like there's a leak of `Pipe` and `->Pipe` types, which matches up. I need to dig into the code and see what I can find.

**Comment by barrucadu** [comment]
A small change here and to [irc-client](https://github.com/barrucadu/irc-client) seems to have largely cleared this up:

```
--- a/Network/IRC/Conduit/Internal/Conduits.hs
+++ b/Network/IRC/Conduit/Internal/Conduits.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE BangPatterns      #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RankNTypes        #-}
 
@@ -7,7 +8,7 @@
 -- License     : MIT
 -- Maintainer  : Michael Walker <mike@barrucadu.co.uk>
 -- Stability   : experimental
--- Portability : OverloadedStrings, RankNTypes
+-- Portability : BangPatterns, OverloadedStrings, RankNTypes
 --
 -- Internal helper conduits. This module is NOT considered to form
 -- part of the public interface of this library.
@@ -25,7 +26,7 @@ import qualified Data.ByteString as B
 chunked :: Monad m => Conduit ByteString m ByteString
 chunked = chunked' ""
   where
-    chunked' leftover = do
+    chunked' !leftover = do
       -- Wait for a value from upstream
       val <- await
```

This surprises me, but did seem to change the retainer set profile:

```
--- a/Network/IRC/Client/Internal.hs
+++ b/Network/IRC/Client/Internal.hs
@@ -118,16 +118,17 @@ forgetful = awaitForever go where
 -- | Block on receiving a message and invoke all matching handlers
 -- concurrently.
 eventSink :: MonadIO m => IRCState s -> Consumer IrcEvent m ()
-eventSink ircstate = await >>= maybe (return ()) (\event -> do
-  let event'  = decodeUtf8 <$> event
-  ignored <- isIgnored ircstate event'
-  unless ignored $ do
-    handlers <- getHandlersFor event' . _eventHandlers <$> getInstanceConfig' ircstate
-    liftIO $ mapM_ (\h -> forkIO $ runReaderT (h event') ircstate) handlers                                                                                                                                           
-
-  -- If disconnected, do not loop.
-  disconnected <- (==Disconnected) <$> getConnState ircstate
-  unless disconnected (eventSink ircstate))
+eventSink ircstate = go where
+  go = await >>= maybe (return ()) (\event -> do
+    let event'  = decodeUtf8 <$> event
+    ignored <- isIgnored ircstate event'
+    unless ignored $ do
+      handlers <- getHandlersFor event' . _eventHandlers <$> getInstanceConfig' ircstate
+      liftIO $ mapM_ (\h -> forkIO $ runReaderT (h event') ircstate) handlers
+
+    -- If disconnected, do not loop.
+    disconnected <- (==Disconnected) <$> getConnState ircstate
+    unless disconnected go)
```

The heap profile still looks a bit iffy (things are creeping up, although more slowly), so I'm not calling this done yet: [hb](https://github.com/barrucadu/irc-conduit/files/533664/hb.pdf), [hc](https://github.com/barrucadu/irc-conduit/files/533665/hc.pdf), [hd](https://github.com/barrucadu/irc-conduit/files/533668/hd.pdf), [hm](https://github.com/barrucadu/irc-conduit/files/533669/hm.pdf), [hr (graph)](https://github.com/barrucadu/irc-conduit/files/533667/hr.pdf), [hr (profile)](https://github.com/barrucadu/irc-conduit/files/533663/hr.txt), [hy](https://github.com/barrucadu/irc-conduit/files/533666/hy.pdf).

**Comment by barrucadu** [comment]
I think this is now fixed in the 0.2.1.1 release. The cause was `exceptionalConduit` leaking on every (whoops) received message. The release also has the stricter `chunked`, which should reduce the overall memory usage a bit too.
